AbstractBlockChain: migrate to CompletableFuture

msgilligan opened the pull request:
This pull request introduces two new utility classes:

* `ListenableCompletionStage` an interface with a single default method that implements Guava `ListenableFuture` using JDK8's `CompletionStage` 
* `ListenableCompletableFuture` a subclass of `CompletableFuture` that implements `ListenableCompletableFuture` (and therefore `ListenableFuture`)

`AbstractBlockchain::getHeightFuture` is updated to return `ListenableCompletableFuture`.

`BlockChainTest` is updated to use `CompletableFuture`  (although it would continue to work without the change)

schildbach commented:
If the other occurences can be migrated as easily as this, I think we can simply migrate everything in one bigger PR.

msgilligan commented:
> If the other occurrences can be migrated as easily as this, I think we can simply migrate everything in one bigger PR.

As I have available time, I will add commits to this PR. Once we've migrating everything and the code looks good we can squash and merge.

schildbach commented:
If you'd like I can take that task. What do you think?

msgilligan commented:
> If you'd like I can take that task. What do you think?

That would be great. I'd be happy to review the PR, of course! Don't know if you can add to this same PR or whether you'd just have to start with this branch, but either way that would be great!

schildbach commented:
Having looked at a few other cases, seems like most migrations will not be as straghforward as the one you picked. So I guess it makes sense to migrate on a case-by-case basis.

schildbach commented:
Looks like we need replacements for the various `Futures` helpers. There is a [nice cheat sheet](https://gist.github.com/kofemann/a721af324f8c35f6e9840f583ca40f05), but it operates on `CompletableFuture` not `ListenableCompletableFuture`.

msgilligan commented:
So merge this one?

schildbach commented:
Hmm, as this affects API (and not just impl detail like #2257 and #2256) I would rather wait until we're sure we don't get stuck with the migration half-way. Atm I'm afraid we might reconsider our strategry outlines in #2254.

msgilligan commented:
> as this affects API  ...I would rather wait until we're sure we don't get stuck with the migration half-way.

Sounds reasonable. (If we want to create multiple PRs for each use case of futures, we could merge the two `utils` methods into `master` so those PRs could be independent.)

msgilligan commented:
@schildbach I just rebased on current master.

Also, I'm trying to figure out how to move forward with this. I'm probably more confident than you that we can convert to `CompletableFuture` relatively easily, but I do understand your concern.

Given that our migration strategy is to use `ListenableCompletableFuture`, I also don't think it would be that bad of an outcome if we decide to release 0.17 without migration all `ListenableFuture` return types.

So, there are a handful of ways we could proceed:

1. Merge this PR as is and then start on some of the other use cases as separate PRs.
2. Split this PR into two PRs: (a) the utility interface + class (b) the change to `AbstractBlockChain`.  We'd then merge (a) and develop PRs for all the other use cases. Once we get PRs for the complete migration we merge them all.
3. Create a branch for the completable future work and develop the entire migration on that branch.

Personally, I prefer #1 as it is the least amount of work (and as I said above I'm confident this path will work out)

schildbach commented:
Ok, I agree. Let's go with (1) and continue on best-effort. I'll merge this in a moment.

schildbach commented:
Merged.
